### ISSUE TITLE
Out of order fields are corrupting data with Insert Into SQL on Spark 3.0

### ISSUE THREAD

**Issue by mehtaashish23** [apache/iceberg#2041]
GIST: https://gist.github.com/mehtaashish23/232dda586040a8e9038b61f3553dde76

While doing INSERT INTO on Iceberg table to append data, the order of fields being commited are wrong, and hence corrupting the table. Checkout the GIST for the code, but if the table has field "f1,f2,f3" and I insert the data using the following code, then it gets the wrong data (out of order fields) into the table. NOTE: dataFrame.write.mode("append").save() works fine and maintains the correct order.

```
sc.parallelize(Seq((s"f3ValueFile2",s"f1ValueFile2",s"f2ValueFile2"))).toDF("f3","f1","f2").registerTempTable("input")
spark.read.format("iceberg").load(masterTablePath).registerTempTable("target")
spark.sql("insert into target (select * from input)").show(false)
```

```
scala> spark.read.format("iceberg").load(masterTablePath).show(false)
+------------+------------+------------+
|f1          |f2          |f3          |
+------------+------------+------------+
|f3ValueFile2|f1ValueFile2|f2ValueFile2|
+------------+------------+------------+
```

**Comment by HeartSaVioR** [comment]
Could you please describe the full context here? Spark version and Iceberg version (exact version including bugfix numbers) are required to let others look into it.

**Comment by RussellSpitzer** [comment]
This is actually Spark's defined behavior for insert into.

https://github.com/apache/spark/blob/271c4f6e00b7bc7c47d84a8e59018e84a19c9822/sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala#L442-L450

```
  * @note Unlike `saveAsTable`, `insertInto` ignores the column names and just uses position-based
   * resolution. For example:   
```

Which I know is confusing, but you'll have to take it up with the Spark folks.

The fix here is to make sure your "INSERT" query matches the column ordering as your base table.

**Comment by HeartSaVioR** [comment]
Yeah, I think @RussellSpitzer is correct on this. (Actually Spark doesn't seem to support INSERT INTO ... SELECT ... natively.) And my quick google says matching by column name during INSERT INTO ... SELECT isn't something other RDBMS would also support.

The safest way to do this is providing explicit column list for both source and target, but Spark doesn't support providing explicit column list in target side, so your only workaround is knowing the order of columns in target table in prior, and ensure your select statement is matching the column list via "explicitly" providing column list instead of simply using "*".

```
spark.sql("insert into target select f1, f2, f3 from input").show(false)
```

This would work. I'd rather say it's safer if Spark supports explicit column list on INSERT INTO, but that's not a matter of Iceberg and you're welcome to ask this to Spark user mailing list. At least as of now I don't see any issue for this.

**Comment by kbendick** [comment]
If `target` is registered in the catalog, fields that need reordering can be determined via a `DESCRIBE` command or via `spark.catalog.listColumns(tableName).select("name").as[String].collect()`

We do this occasionally at my work. I can try and take a look to see if there's a cleaner approach that we might use.

```scala
val targetColumns = spark.sql("DESCRIBE EXTENDED target").select("col_name").as[String].collect()
spark.sql(s"insert into target select ${targetColumns.mkString(", ")} from input")
```

**Comment by dongjoon-hyun** [comment]
As @HeartSaVioR mentioned, this is not a bug. Instead, it's the standard SQL behavior when there is no explicit column list.
```
postgres=# create table t as select 1 a, 2 b;
SELECT 1
postgres=# select * from t;
 a | b 
---+---
 1 | 2
(1 row)

postgres=# create table s(b int, a int);
CREATE TABLE
postgres=# insert into s (select * from t);
INSERT 0 1
postgres=# select * from s;
 b | a 
---+---
 1 | 2
(1 row)
```

Apache Spark 3.1.1 will start to support the explicit column list in `INSERT INTO` statement.
- https://issues.apache.org/jira/browse/SPARK-32976

Although Spark 3.1.1 is still RC1 and Iceberg 0.11.0 RC0 is not compatible with it yet, I believe this issue will be resolved by Spark 3.1.1 and Iceberg 0.11.1 combination soon.

**Comment by HeartSaVioR** [comment]
Oh great to know it is addressed in Spark via SPARK-32976 and will be available. Thanks for the info!
